streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorMetadataTest.java

```diff
@@ -0,0 +1,151 @@
+        final long value1 = 1L, value2 = 2L, value3 = 3L;
+
+        metadata.addMetadata(key1, value1);
+        metadata.addMetadata(key2, value2);
+        metadata.addMetadata(key3, value3);
+
+        final byte[] serialized = metadata.serialize();
+        final ProcessorMetadata deserialized = ProcessorMetadata.deserialize(serialized);
+
+        assertThat(deserialized.getMetadata(key1), is(value1));
+        assertThat(deserialized.getMetadata(key2), is(value2));
+        assertThat(deserialized.getMetadata(key3), is(value3));
+    }
+
+    @Test
+    public void shouldDeserializeNull() {
+        final ProcessorMetadata deserialized = ProcessorMetadata.deserialize(null);
+        assertThat(deserialized, is(ProcessorMetadata.emptyMetadata()));
+    }
+
+    @Test
+    public void shouldUpdate() {
+        final ProcessorMetadata emptyMeta = ProcessorMetadata.emptyMetadata();
+        emptyMeta.update(null);
+
+        assertThat(emptyMeta, is(ProcessorMetadata.emptyMetadata()));
+
+        {
+            final Map<String, Long> map1 = new HashMap<>();
+            map1.put("key1", 1L);
+            map1.put("key2", 2L);
+            final ProcessorMetadata metadata1 = ProcessorMetadata.with(map1);
+            emptyMeta.update(metadata1);
+            assertThat(emptyMeta.getMetadata("key1"), is(1L));
+            assertThat(emptyMeta.getMetadata("key2"), is(2L));
+        }
+
+        {
+            final Map<String, Long> map1 = new HashMap<>();
+            map1.put("key1", 0L);
+            map1.put("key2", 1L);
+            final ProcessorMetadata metadata1 = ProcessorMetadata.with(map1);
+            emptyMeta.update(metadata1);
+            assertThat(emptyMeta.getMetadata("key1"), is(1L));
+            assertThat(emptyMeta.getMetadata("key2"), is(2L));
+        }
+
+        {
+            final Map<String, Long> map1 = new HashMap<>();
+            map1.put("key1", 2L);
+            map1.put("key2", 3L);
+            final ProcessorMetadata metadata1 = ProcessorMetadata.with(map1);
+            emptyMeta.update(metadata1);
+            assertThat(emptyMeta.getMetadata("key1"), is(2L));
+            assertThat(emptyMeta.getMetadata("key2"), is(3L));
+        }
+    }
+
+    @Test
+    public void shouldUpdateCommitFlag() {
+        final ProcessorMetadata emptyMeta = ProcessorMetadata.emptyMetadata();
+        assertFalse(emptyMeta.needCommit());
+
+        emptyMeta.setNeedsCommit(true);
+        assertTrue(emptyMeta.needCommit());
+
+        emptyMeta.setNeedsCommit(false);
+        assertFalse(emptyMeta.needCommit());
+
+        emptyMeta.addMetadata("key1", 1L);
+        assertTrue(emptyMeta.needCommit());
+
+        final Map<String, Long> map1 = new HashMap<>();
+        map1.put("key1", 2L);
+        map1.put("key2", 3L);
+        final ProcessorMetadata metadata1 = ProcessorMetadata.with(map1);
+        emptyMeta.update(metadata1);
+        assertTrue(emptyMeta.needCommit());
+    }
+}
```

Comment:
Should we add a test that the commit flag has no impact on `equals` and `hashCode` ?